### ISSUE TITLE
Function annotation not hoisting with function declarations

### ISSUE THREAD

**Issue by kentcdodds** [olov/ng-annotate#168]
Given this valid JavaScript:

```javascript
// example.js
function foo() {
  $injector.invoke(someFunction);
  
  // @ngInject
  function someFunction(someService) {
    someService.setSomething('bar');
  }
}
```

The result of `ng-annotate -a example.js` is

```javascript
// example.js
function foo() {
  $injector.invoke(someFunction);
  
  // @ngInject
  function someFunction(someService) {
    someService.setSomething('bar');
  }
  someFunction.$inject = ["someService"];
}
```

The problem is, in strict di mode, this would throw an error because the `$inject` property gets set *after* the `$injector.invoke` happens.

To fix this, I would expect the `$inject` assignment to be hoisted where the function declaration is hoisted (to the top of the closure). So output like this would be correct (and fix the issue):

```javascript
// example.js
function foo() {
  someFunction.$inject = ["someService"];
  $injector.invoke(someFunction);
  
  // @ngInject
  function someFunction(someService) {
    someService.setSomething('bar');
  }
}
```

Note, this behavior should only apply to function declarations, not function expressions. So if instead of `function someFunction(someService)...` it was `var someFunction = function(someService)...` then the current output would be correct.

**Comment by olov** [comment]
Thanks for the clear examples. I just pushed out 1.0.0 and hope to have a fix for this issue in 1.1. Push `someFunction` to the top as a temporary workaround until then.

**Comment by kentcdodds** [comment]
Thanks, looking forward to this update. I write all my code [like this](https://medium.com/@kentcdodds/angular-newspaper-code-structure-19eee9e90e45), and most of the time it's not a problem. Looking forward to the fix :-)

**Comment by olov** [comment]
Ok, so I've created a prototype that addresses this by always putting the `foo.$inject = [...]` statement at the beginning of the function/script (but after any directive prologues). This only happens for *function declarations* so function expressions stays as-is. On the plus-side, this fixes the issue. On the minus-side, since the function and its annotation is now apart, it makes it harder to inspect the output.

I did not think this issue caused much concern in practice. I thought that for the rare cases when someone was hit by it, it's easily handled with an explicit `"ngInject";`. Then I saw https://github.com/stackus/manageiq/commit/41b7ed7d40ee27b07aaf35d7b0ac0887449b8710.

I'm thinking that I'll go ahead with this change for correctness and predictability reasons. Feedback? cc @omsmith @mzgol
